### PR TITLE
fs/directory: Using file api to implement operations on directories

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
1. Using file api to implement operations on directories.
```
opendir -> open with O_DIRECOTRY
closedir -> close
rewindir-> lseek
readdir -> read
seekdir -> lseek
telldir -> lseek
```
2. Using user memory to obtain directory entries and avoid using kernel space buffers. Therefore, add the entry member in DIR and avoid using the fd_dir member in `fs_dirent_s`.
3. Removing the private information of all file systems in `fs_dirent_s` and use opendir to create them. 
4. In order to better realize the “*at” system calls api, such as openat, https://linux.die.net/man/2/openat etc.

## Impact
Change opendir and readdir prototypes:
```
int     (*opendir)(FAR struct inode *mountpt, FAR const char *relpath,FAR struct fs_dirent_s **dir);
int     (*readdir)(FAR struct inode *mountpt, FAR struct fs_dirent_s *dir, FAR struct dirent *entry);
```

## Testing
Local test

some discussion here: https://github.com/apache/incubator-nuttx/pull/4300

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@yamt this is the first patchset to support openat.

File: arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c [hunk 1]
```diff
@@ -749,7 +750,8 @@ static int cxd56_powermgr_procfs_closedir(struct fs_dirent_s *dir)
  *
  ****************************************************************************/
 
-static int cxd56_powermgr_procfs_readdir(struct fs_dirent_s *dir)
+static int cxd56_powermgr_procfs_readdir(FAR struct fs_dirent_s *dir,
+                                         FAR struct dirent *entry)
```
Comment (xiaoxiang781216) [inline arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c:744]: remove FAR from both

File: arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c [hunk 2]
```diff
@@ -781,7 +783,7 @@ static int cxd56_powermgr_procfs_readdir(struct fs_dirent_s *dir)
  *
  ****************************************************************************/
 
-static int cxd56_powermgr_procfs_rewinddir(struct fs_dirent_s *dir)
+static int cxd56_powermgr_procfs_rewinddir(FAR struct fs_dirent_s *dir)
```
Comment (xiaoxiang781216) [inline arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c:776]: remove FAR

File: arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c [hunk 3]
```diff
@@ -711,7 +710,7 @@ static int cxd56_powermgr_procfs_opendir(const char *relpath,
     }
 
   procfs->index = 0;
-  dir->u.procfs = (void *)procfs;
+  *dir = (FAR struct fs_dirent_s *)procfs;
```
Comment (xiaoxiang781216) [inline arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c:713]: remove FAR

File: arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c [hunk 4]
```diff
@@ -755,9 +745,9 @@ static int cxd56_powermgr_procfs_readdir(FAR struct fs_dirent_s *dir,
 {
   struct cxd56_powermgr_procfs_dir_s *procfs;
 
-  DEBUGASSERT(dir && dir->u.procfs);
+  DEBUGASSERT(dir);
 
-  procfs = (struct cxd56_powermgr_procfs_dir_s *)dir->u.procfs;
+  procfs = (FAR struct cxd56_powermgr_procfs_dir_s *)dir;
```
Comment (xiaoxiang781216) [inline arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c:750]: remove FAR

File: arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c [hunk 5]
```diff
@@ -787,9 +777,9 @@ static int cxd56_powermgr_procfs_rewinddir(FAR struct fs_dirent_s *dir)
...
```
Comment (xiaoxiang781216) [inline arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c:782]: ditto

Comment (Donny9) [reply arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c]: Done!

File: arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c (hunk 3 above)
Comment (Donny9) [reply arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c]: Done

File: arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c (hunk 2 above)
Comment (Donny9) [reply arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c]: Done.

File: arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c (hunk 1 above)
Comment (Donny9) [reply arch/arm/src/cxd56xx/cxd56_powermgr_procfs.c]: Done.

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by davids5 — CHANGES_REQUESTED** [review]
@Donny9, @xiaoxiang781216  - 
What was the motivation for this change?

For a change of this size nature all the file system will need to tested and result posted as well.  the statistics on the impact on memory ram/flash.

**Comment by xiaoxiang781216** [comment]
> @Donny9, @xiaoxiang781216 - What was the motivation for this change?
> 

1. From the design perspective, directory is also a file(https://en.wikipedia.org/wiki/Everything_is_a_file).
2. Support dirfd(https://pubs.opengroup.org/onlinepubs/9699919799/functions/dirfd.html)

> For a change of this size nature all the file system will need to tested and result posted as well. the statistics on the impact on memory ram/flash.

@Donny9 please give a summary.

File: fs/binfs/fs_binfs.c [hunk 6]
```diff
@@ -63,9 +72,12 @@ static int     binfs_fstat(FAR const struct file *filep,
                            FAR struct stat *buf);
 
 static int     binfs_opendir(struct inode *mountpt, const char *relpath,
```
Comment (pkarashchenko) [inline fs/binfs/fs_binfs.c:74]:
```suggestion
static int     binfs_opendir(FAR struct inode *mountpt, FAR const char *relpath,
```

File: fs/cromfs/fs_cromfs.c [hunk 7]
```diff
@@ -1356,9 +1392,12 @@ static int cromfs_readdir(struct inode *mountpt, struct fs_dirent_s *dir)
 
 static int cromfs_rewinddir(struct inode *mountpt, struct fs_dirent_s *dir)
```
Comment (pkarashchenko) [inline fs/cromfs/fs_cromfs.c:1393]:
```suggestion
static int cromfs_rewinddir(FAR struct inode *mountpt, FAR struct fs_dirent_s *dir)
```

File: fs/fat/fs_fat32.h [hunk 8]
```diff
@@ -1054,8 +1067,9 @@ EXTERN int    fat_allocatedirentry(struct fat_mountpt_s *fs,
                                    struct fat_dirinfo_s *dirinfo);
 EXTERN int    fat_freedirentry(struct fat_mountpt_s *fs,
                                struct fat_dirseq_s *seq);
```
Comment (pkarashchenko) [inline fs/fat/fs_fat32.h:1069]:
```suggestion
EXTERN int    fat_freedirentry(FAR struct fat_mountpt_s *fs,
                               FAR struct fat_dirseq_s *seq);
```

File: fs/fat/fs_fat32.h [hunk 9]
```diff
@@ -1054,8 +1067,9 @@ EXTERN int    fat_allocatedirentry(struct fat_mountpt_s *fs,
                                    struct fat_dirinfo_s *dirinfo);
```
Comment (pkarashchenko) [inline fs/fat/fs_fat32.h:1067]:
```suggestion
                                   FAR struct fat_dirinfo_s *dirinfo);
```

File: fs/vfs/fs_dir.c [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+#endif
+};
+
+static struct inode g_dir_inode =
+{
+  .i_crefs = 1,
+  .u =
+    {
+      .i_ops = &g_dir_fileops,
+    },
```
Comment (pkarashchenko) [inline fs/vfs/fs_dir.c:86]: can we move to C89 initializer?

File: libs/libc/dirent/lib_dirfd.c [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+ *   On error, -1 is returned, and errno is set to indicate
+ *   the cause of the error.
+ *
+ *   EINVAL - dirp does not refer to a valid directory stream.
+ *
+ ****************************************************************************/
+
+int dirfd(FAR DIR *dirp)
+{
+  if (dirp)
```
Comment (pkarashchenko) [inline libs/libc/dirent/lib_dirfd.c:58]:
```suggestion
  if (dirp != NULL)
```

File: libs/libc/dirent/lib_rewinddir.c [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+ *     call to opendir();
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void rewinddir(FAR DIR *dirp)
+{
+  if (dirp)
```
Comment (pkarashchenko) [inline libs/libc/dirent/lib_rewinddir.c:55]:
```suggestion
  if (dirp != NULL)
```

File: libs/libc/dirent/lib_seekdir.c [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+ *   offset -- offset to seek to
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void seekdir(FAR DIR *dirp, off_t offset)
+{
+  if (dirp)
```
Comment (pkarashchenko) [inline libs/libc/dirent/lib_seekdir.c:57]:
```suggestion
  if (dirp != NULL)
```

File: libs/libc/dirent/lib_telldir.c [hunk 14]
```diff
@@ -61,15 +56,11 @@
 
 off_t telldir(FAR DIR *dirp)
 {
-  struct fs_dirent_s *idir = (struct fs_dirent_s *)dirp;
-
-  if (!idir || !idir->fd_root)
+  if (dirp)
```
Comment (pkarashchenko) [inline libs/libc/dirent/lib_telldir.c:59]:
```suggestion
  if (dirp != NULL)
```

Comment (Donny9) [reply libs/libc/dirent/lib_telldir.c:59]: Done!

File: libs/libc/dirent/lib_seekdir.c (hunk 13 above)
Comment (Donny9) [reply libs/libc/dirent/lib_seekdir.c:57]: Done!

File: libs/libc/dirent/lib_rewinddir.c (hunk 12 above)
Comment (Donny9) [reply libs/libc/dirent/lib_rewinddir.c:55]: Done!

File: libs/libc/dirent/lib_dirfd.c (hunk 11 above)
Comment (Donny9) [reply libs/libc/dirent/lib_dirfd.c:58]: Done!

File: fs/fat/fs_fat32.h (hunk 8 above)
Comment (Donny9) [reply fs/fat/fs_fat32.h:1069]: keep original?

File: fs/fat/fs_fat32.h (hunk 9 above)
Comment (Donny9) [reply fs/fat/fs_fat32.h:1067]: keep original?

File: fs/cromfs/fs_cromfs.c (hunk 7 above)
Comment (Donny9) [reply fs/cromfs/fs_cromfs.c:1393]: Done

File: fs/binfs/fs_binfs.c (hunk 6 above)
Comment (Donny9) [reply fs/binfs/fs_binfs.c:74]: Done

File: fs/binfs/fs_binfs.c [hunk 15]
```diff
@@ -277,8 +289,10 @@ static int binfs_fstat(FAR const struct file *filep, FAR struct stat *buf)
  ****************************************************************************/
 
 static int binfs_opendir(struct inode *mountpt, const char *relpath,
```
[the remark this answers is not in the document]
Comment (Donny9) [reply fs/binfs/fs_binfs.c:291]: Done

File: fs/binfs/fs_binfs.c [hunk 16]
```diff
@@ -277,8 +289,10 @@ static int binfs_fstat(FAR const struct file *filep, FAR struct stat *buf)
...
-                         struct fs_dirent_s *dir)
+                         struct fs_dirent_s **dir)
```
[the remark this answers is not in the document]
Comment (Donny9) [reply fs/binfs/fs_binfs.c:292]: Done

File: fs/binfs/fs_binfs.c [hunk 17]
```diff
@@ -63,9 +72,12 @@ static int     binfs_fstat(FAR const struct file *filep,
                            FAR struct stat *buf);
 
 static int     binfs_opendir(struct inode *mountpt, const char *relpath,
-                             struct fs_dirent_s *dir);
+                             struct fs_dirent_s **dir);
```
[the remark this answers is not in the document]
Comment (Donny9) [reply fs/binfs/fs_binfs.c:75]: Done

File: fs/vfs/fs_dir.c (hunk 10 above)
Comment (Donny9) [reply fs/vfs/fs_dir.c:86]: what should i do?

File: libs/libc/dirent/lib_telldir.c (hunk 14 above)
Comment (Donny9) [reply libs/libc/dirent/lib_telldir.c:59]: > What was the motivation for this change?
> 
> For a change of this size nature all the file system will need to tested and result posted as well. the statistics on the impact on memory ram/flash.

@davids5 
Statistics for three situation: nodir, dir, dir_with_fchdir_dirfd
```
djz:nuttx$ ls -l vela_cp.bin 
-rwxrwxr-x 1 djz djz 2066272 8月   9 10:36 vela_cp.bin

djz:nuttx$ ls -l vela_cp_dir.bin 
-rwxrwxr-x 1 djz djz 2066260 8月   9 10:56 vela_cp_dir.bin

djz:nuttx$ ls -l vela_cp_dir_with_fchdir_dirfd.bin 
-rwxrwxr-x 1 djz djz 2066428 8月   9 11:10 vela_cp_dir_with_fchdir_dirfd.bin

no_dir:
cp> free
                   total       used       free    largest  nused  nfree
        Umem:     519096     210392     308704     304384    672      8
dir:
cp> free
                   total       used       free    largest  nused  nfree
        Umem:     519032     210392     308640     304864    672     10
dir_with_fchdir_dirfd:
cp> free
                   total       used       free    largest  nused  nfree
        Umem:     519096     210392     308704     304928    672      8
        
filesystem usage：
no_dir:
cp> df -h
  Filesystem    Size      Used  Available Mounted on
  rpmsgfs      2298M      121M      2177M /data
  romfs         512B      512B         0B /etc
  littlefs      256K        8K       248K /misc
  procfs          0B        0B         0B /proc
  tmpfs         512B      512B         0B /tmp
  rpmsgfs      1718K     1718K         0B /vendor

dir:
cp> df -h
  Filesystem    Size      Used  Available Mounted on
  rpmsgfs      2298M      121M      2177M /data
  romfs         512B      512B         0B /etc
  littlefs      256K        8K       248K /misc
  procfs          0B        0B         0B /proc
  tmpfs         512B      512B         0B /tmp
  rpmsgfs      1718K     1718K         0B /vendor

dir_with_fchdir_dirfd:
cp> df -h
  Filesystem    Size      Used  Available Mounted on
  rpmsgfs      2298M      122M      2176M /data
  romfs         512B      512B         0B /etc
  littlefs      256K        8K       248K /misc
  procfs          0B        0B         0B /proc
  tmpfs         512B      512B         0B /tmp
  rpmsgfs      1718K     1718K         0B /vendor
```

**Comment by Donny9** [comment]
> @Donny9, @xiaoxiang781216 - What was the motivation for this change?
> 
> For a change of this size nature all the file system will need to tested and result posted as well. the statistics on the impact on memory ram/flash.

@davids5 
Tree situation: nodir, dir, dir_with_fchdir_dirfd
```
djz:nuttx$ ls -l vela_cp.bin 
-rwxrwxr-x 1 djz djz 2066272 8月   9 10:36 vela_cp.bin

djz:nuttx$ ls -l vela_cp_dir.bin 
-rwxrwxr-x 1 djz djz 2066260 8月   9 10:56 vela_cp_dir.bin

djz:nuttx$ ls -l vela_cp_dir_with_fchdir_dirfd.bin 
-rwxrwxr-x 1 djz djz 2066428 8月   9 11:10 vela_cp_dir_with_fchdir_dirfd.bin

no_dir:
cp> free
                   total       used       free    largest  nused  nfree
        Umem:     519096     210392     308704     304384    672      8
dir:
cp> free
                   total       used       free    largest  nused  nfree
        Umem:     519032     210392     308640     304864    672     10
dir_with_fchdir_dirfd:
cp> free
                   total       used       free    largest  nused  nfree
        Umem:     519096     210392     308704     304928    672      8
        
filesystem usage：
no_dir:
cp> df -h
  Filesystem    Size      Used  Available Mounted on
  rpmsgfs      2298M      121M      2177M /data
  romfs         512B      512B         0B /etc
  littlefs      256K        8K       248K /misc
  procfs          0B        0B         0B /proc
  tmpfs         512B      512B         0B /tmp
  rpmsgfs      1718K     1718K         0B /vendor

dir:
cp> df -h
  Filesystem    Size      Used  Available Mounted on
  rpmsgfs      2298M      121M      2177M /data
  romfs         512B      512B         0B /etc
  littlefs      256K        8K       248K /misc
  procfs          0B        0B         0B /proc
  tmpfs         512B      512B         0B /tmp
  rpmsgfs      1718K     1718K         0B /vendor

dir_with_fchdir_dirfd:
cp> df -h
  Filesystem    Size      Used  Available Mounted on
  rpmsgfs      2298M      122M      2176M /data
  romfs         512B      512B         0B /etc
  littlefs      256K        8K       248K /misc
  procfs          0B        0B         0B /proc
  tmpfs         512B      512B         0B /tmp
  rpmsgfs      1718K     1718K         0B /vendor
```

**Comment by xiaoxiang781216** [comment]
@davids5 please take a look, the change is minor.

**Review by davids5 — APPROVED** [review]
@Donny9 - thank you for the information. It looks good. Please fix the C89 compatibility issue.

File: fs/vfs/fs_dir.c (hunk 10 above)
Comment (Donny9) [reply fs/vfs/fs_dir.c]: Done!

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by PetervdPerk-NXP** [comment]
opendir(), readerdir() etc are POSIX.1-2001 functions.
This PR removes these functions calls hence breaking with the POSIX spec.
Can't have fallback functions to still stay compatible with POSIX.1-2001?

**Comment by btashton** [comment]
FYI a report from @lupyuen that this change may have broken some out of tree configs

https://twitter.com/MisterTechBlog/status/1560839031990104064

@lupyuen can you give some more context.

**Comment by lupyuen** [comment]
Thanks @btashton! PineDio Stack BL604 fails my LoRaWAN Network Test because it couldn't receive any LoRaWAN Packets (Join Accept Response). Sending LoRaWAN Packets is OK. I traced it to this commit: https://github.com/apache/incubator-nuttx/commit/4dbf4555ebb743a908e3016d43b5b456d7a2cf16

The LoRaWAN Test calls my [LoRaWAN Library](https://github.com/lupyuen/LoRaMac-node-nuttx), which calls my [Semtech SX1262 Library](https://github.com/lupyuen/lora-sx1262/tree/lorawan), which calls the [NimBLE Porting Layer](https://github.com/lupyuen/nimble-porting-nuttx) (for multithreading functions).

Since the commit https://github.com/apache/incubator-nuttx/commit/4dbf4555ebb743a908e3016d43b5b456d7a2cf16 has to do with File System, maybe something inside NimBLE Porting Layer broke.

I plan to trace the calls to [`dir_open`](https://github.com/apache/incubator-nuttx/blob/master/fs/vfs/fs_dir.c#L397-L402), [`dir_ioctl`](https://github.com/apache/incubator-nuttx/blob/master/fs/vfs/fs_dir.c#L548-L560) and [`dir_allocate`](https://github.com/apache/incubator-nuttx/blob/master/fs/vfs/fs_dir.c#L566-L608) to see what changed.

The weird thing about this bug: LoRaWAN works perfectly fine on PineCone BL602, only PineDio Stack BL604 is affected. (Maybe due to the different drivers and apps loaded in memory?)

**Comment by xiaoxiang781216** [comment]
@lupyuen it's strange that dir_open/dir_ioctl can generate this side effect since:

1.  dir_allocate is already used in open(O_DIRECTORY)
2. dir_open/dir_ioctl is used to implement fchrdir which don't have any caller yet

**Comment by lupyuen** [comment]
Hi @xiaoxiang781216 after I added logging for `dir_open` / `dir_ioctl` / `dir_close`, LoRaWAN worked OK, which is really weird.

You're right, `dir_open` / `dir_ioctl` / `dir_close` are not called, according to the log. So I don't think they are the cause of the problem.

I'll dig deeper inside the GPIO Interrupt Handler and NimBLE Porting Layer to find out why LoRaWAN isn't handling the Packet Received Interrupt from the SX1262 Radio Module. Thanks!

**Pull request merged**